File: pilha/pilha.c
#include <stdio.h>
#include <stdlib.h>

// Definindo o tamanho máximo da pilha
#define tamanhoMaximo 100

/* Criando a pilha como uma lista
   com tamanho 100 (tamanhoMaximo) */
int pilha[tamanhoMaximo];

/* Variável para acompanhar o topo da pilha, 
iniciada com -1 para indicar que a pilha está vazia*/
int top = -1;

// Funcao para verificar se a pilha está vazia
int estaVazia();

// Funcao para verificar se a pilha está cheia
int estaCheia();

// Funcao para adicionar um elementa a pilha (push)
void adicionarElemento(int);

// Funcao para remover um elemento da pilha (pop)
int removerElemento();

// Funcao para olhar o topo da pilha (peek)
int verTopoDaPilha();

// Funcao Principal (main), a que será executada primeiro
int main()
{
    // Adicionando elementos à pilha
    adicionarElemento(1); // pilha: [1]
    adicionarElemento(2); // pilha: [1, 2]
    adicionarElemento(3); // pilha: [1, 2, 3]

    // Iterando sobre os elementos da pilha para imprimir sua representação
    // Resultado: [1, 2, 3]
    printf("[");
    for (int i = 0; i <= top; i++)
    {
        if (i == top)
            printf("%d]\n", pilha[i]);
        else
            printf("%d, ", pilha[i]);
    }

    printf("Elemento removido: %d\n", removerElemento());  // Deve imprimir 3, Pilha agora: [1, 2]

    printf("Topo da pilha: %d\n", verTopoDaPilha());  // Deve imprimir 2

    printf("A pilha está vazia? %s\n", estaVazia() ? "Sim" : "Não");  // Deve imprimir "Não"

    removerElemento();  // Remove 2, Pilha: [1]
    removerElemento();  // Remove 1, Pilha: []

    printf("A pilha está vazia? %s\n", estaVazia() ? "Sim" : "Não");  // Deve imprimir "Sim"

    return 0;
}

int estaVazia()
{
    return top == -1;
}

int estaCheia()
{
    return top == tamanhoMaximo -1;
}

void adicionarElemento(int elemento)
{
    // Verificar se a pilha está cheia antes de adicionar um elemento
    if (estaCheia())
    {
        printf("A pilha está cheia. Não é possível adicionar mais elementos.\n");
        return;
    }

    // Incrementar o índice do topo da pilha (top) antes de adicionar o elemento
    // Isso garante que o novo elemento seja adicionado no próximo espaço disponível na pilha
    // Se a pilha estiver vazia, top será -1. Portanto, top + 1 será o índice 0.
    // Se a pilha tiver elementos, top + 1 será o próximo espaço disponível após o último elemento.
    top++;  // Incrementa o índice do topo da pilha
    pilha[top] = elemento;  // Adiciona o elemento ao topo da pilha
}

int removerElemento()
{
    // Verificar se a pilha está vazia antes de remover um elemento
    if (estaVazia())
    {
        printf("A pilha está vazia. Não é possível remover elementos.\n");
        return -1;  // Retorna um valor indicando erro, já que a pilha está vazia
    }

    // Retornar e remover o elemento do topo da pilha
    // O operador '--top' primeiro usa o valor de 'top' e depois o decrementa,
    // portanto, 'pilha[top--]' retorna o elemento do topo da pilha e, em seguida,
    // decrementa o índice do topo para apontar para o próximo elemento da pilha.
    return pilha[top--];
}


int verTopoDaPilha()
{
    if (estaVazia())
    {
        printf("A pilha está vazia.\n");
        return -1;
    }
    return pilha[top];
}

File: pilha/pilha.py
# Criando uma pilha usando uma lista em Python. 
# As listas em Python têm métodos que nos permitem simular o comportamento de uma pilha.

# Inicializamos uma lista vazia para representar a pilha
pilha = []

# Operação de empilhar (push): Adiciona um elemento ao topo da pilha.  
pilha.append(1)  # Pilha: [1]
pilha.append(2)  # Pilha: [1, 2]
pilha.append(3)  # Pilha: [1, 2, 3]

# Mostrando a pilha
# Resultado: [1, 2, 3]
print("Essa é a pilha:", pilha)

# Operação de desempilhar (pop): Remove e retorna o elemento do topo da pilha.
# Usamos o método pop() da lista para isso.
elemento_removido = pilha.pop()  # elemento_removido = 3, Pilha agora é [1, 2]
print("Esse é o elemento removido:", elemento_removido)
# Mostrando a pilha
# Resultado: [1, 2]
print("Essa é a pilha:", pilha)

# Operação de olhar o topo da pilha: Apenas verifica o elemento do topo sem removê-lo.
# Para isso, acessamos o último elemento da lista.
# Usamos -1 para acessar o último elemento da lista.
topo_da_pilha = pilha[-1]  # topo_da_pilha = 2
print("Esse é o topo da pilha:", topo_da_pilha)

# Verificar se a pilha está vazia: Podemos checar se o comprimento da lista é zero.
esta_vazia = len(pilha) == 0  # esta_vazia = False (pois a pilha tem elementos)


# Empilhando mais elementos
pilha.append(4)  # Pilha: [1, 2, 4]
pilha.append(5)  # Pilha: [1, 2, 4, 5]

print(pilha)

# Desempilhando mais elementos
elemento_removido = pilha.pop()  # elemento_removido = 5, Pilha agora é [1, 2, 4]
elemento_removido = pilha.pop()  # elemento_removido = 4, Pilha agora é [1, 2]
print(pilha)

# Verificando o topo novamente
topo_da_pilha = pilha[-1]  # topo_da_pilha = 2

# Verificando se a pilha está vazia novamente
esta_vazia = len(pilha) == 0  # esta_vazia = False

# Desempilhando todos os elementos restantes
elemento_removido = pilha.pop()  # elemento_removido = 2, Pilha agora é [1]
elemento_removido = pilha.pop()  # elemento_removido = 1, Pilha agora é []

# Verificando se a pilha está vazia depois de desempilhar tudo
esta_vazia = len(pilha) == 0  # esta_vazia = True
